docs/source/types.rst

```diff
@@ -0,0 +1,231 @@
+
+.. math::
+    
+    t = t_0 + t_1 \epsilon_1 + t_2 \epsilon_2 + t_3 \epsilon_3 + t_4 \epsilon_1 \epsilon_2 + t_5 \epsilon_1 \epsilon_3 + t_6 \epsilon_2 \epsilon_3 + t_7 \epsilon_1 \epsilon_2 \epsilon_3
+
+The ``TensorNumber`` type stores an ensemble of such numbers. The  type's ``tens`` field is used to store the coefficients of the :math:`\epsilon_1 \epsilon_2 \epsilon_3` terms (:math:`t_7` above). All the other components are contained in the ``hessnum`` field.
+
+The action of a univariate function :math:`f` on an individual :math:`3^{\text{rd}}`-order hyper-dual number is defined as:
+
+.. math::
+    
+    f(t) = & f(t_0) + f'(t_0) \cdot (t_1 ϵ_1 + t_2 ϵ_2 + t_3 ϵ_3 + t_4 ϵ_1 ϵ_2 + t_5 ϵ_1 ϵ_3 + t_6 ϵ_2 ϵ_3 + t_7 ϵ_1 ϵ_2 ϵ_3) + \\ 
+           & f''(t_0) \cdot (t_1 t_2 ϵ_2 ϵ_1 + t_1 t_3 ϵ_3 ϵ₁ + t_3 t_4 ϵ_2 ϵ_3 ϵ_1 + t_2 t_5 ϵ_2 ϵ_3 ϵ_1 + t_1 t_6 ϵ_2 ϵ_3 ϵ_1 + t_2 t_3 ϵ_2 ϵ_3) + \\
+           & f'''(t_0) \cdot (t_1 t_2 t_3 ϵ_2 ϵ_3 ϵ_1)
+
+We can rearrange this to get the new coefficient of the :math:`\epsilon_1 \epsilon_2 \epsilon_3` term:
+
+
+.. math::
+
+    f(t)_{\epsilon_1 \epsilon_2 \epsilon_3} = f'(t_0) \cdot t_7 + f''(t_0) \cdot (t_3 t_4 + t_2 t_5 + t_1 t_6) + f'''(t_0) \cdot t_1 t_2 t_3
+
+
+As was said, the ``TensorNumber`` type is basically a :math:`3^{\text{rd}}`-order extension of the previous types - it's an ensemble of hyper-dual numbers, and is implemented such that the lower-order partial values are reused where possible. Thus, one could draw the same kind of tree representation for a ``TensorNumber`` instance that is drawn above for the ``HessianNumber`` and ``GradientNumber`` types (though even for :math:`N=4`, it's too large for us to show here).
+
+Relating the mathematics to the implementation, here's what the indexing structure looks like for an individual number in the ``TensorNumber`` ensemble (where the :math:`a`, :math:`b`, and :math:`c` values are stored in ``t.hessnum``, while the :math:`d` values are stored in ``t.tens``):
+
+.. math::
+    
+    t_{N_{ijk}} = a + b_i \epsilon_1 + b_j \epsilon_2 + b_k \epsilon_3 + c_{q_{ij}} \epsilon_1 \epsilon_2 + c_{q_{ik}} \epsilon_1 \epsilon_3 + c_{q_{jk}} \epsilon_2 \epsilon_3 + d_{p_{ijkN}} \epsilon_1 \epsilon_2 \epsilon_3
+
+where
+
+.. math::
+    
+    1 \leq i \leq N
+
+    i \leq j \leq N
+
+    i \leq k \leq j
+
+    M = \frac{N(N+1)(N+2)}{6}
+
+    d_p \in \{d_1, d_2, d_3...d_{M-1}, d_M\}
+
+    p_{ijkN} &= k + \left[\sum_{\alpha=i}^{j}\alpha-i+1\right] + \left[\sum_{\alpha=1}^{i-1} \sum_{\beta=\alpha}^{N}\beta-\alpha+1\right] \\
```

Comment:
This line appears to be incorrect. It's supposed to be formula for going from (i,j,k) --> p in the loop

```julia
function loop(N)
    p = 1
    for i in 1:N
        for j in i:N
            for k in i:j
                println((i,j,k), "-->", p)
                p += 1
            end
        end
    end
end
```

Must've made a math error somewhere...